File: tools.py
"""Custom MCP tools exposed to agents via an in-process server.

The factory :func:`create_harness_tools` returns a ``McpSdkServerConfig``
that can be passed to ``ClaudeCodeOptions.mcp_servers``.
"""

from __future__ import annotations

import json
import os
from datetime import datetime, timezone
from typing import Any

from claude_code_sdk import SdkMcpTool, create_sdk_mcp_server

# ---------------------------------------------------------------------------
# JSON schemas used by the ``validate_json`` tool
# ---------------------------------------------------------------------------

_REQUIRED_KEYS: dict[str, list[str]] = {
    "product_spec": ["name", "description", "design_system", "sprints", "stack"],
    "feature_list_item": ["id", "sprint", "category", "description", "passes"],
    "sprint_contract": ["sprint_id", "sprint_name", "agreed_deliverables"],
    "qa_report": ["sprint_id", "overall_score", "verdict", "scores", "bugs"],
    "contract_review": ["sprint_id", "approved", "feedback"],
}


# ---------------------------------------------------------------------------
# Tool handlers (defined at module level, workspace_dir injected later)
# ---------------------------------------------------------------------------

def _make_update_progress(workspace_dir: str):
    """Return an ``update_progress`` handler bound to *workspace_dir*."""

    async def handler(args: dict[str, Any]) -> dict[str, Any]:
        phase = args.get("phase", "unknown")
        message = args.get("message", "")
        path = os.path.join(workspace_dir, "claude-progress.txt")
        ts = datetime.now(timezone.utc).strftime("%Y-%m-%d %H:%M:%S UTC")
        line = f"[{ts}] [{phase}] {message}\n"
        with open(path, "a", encoding="utf-8") as fh:
            fh.write(line)
        return {"content": [{"type": "text", "text": f"Progress updated: {line.strip()}"}]}

    return handler


def _make_validate_json(workspace_dir: str):
    """Return a ``validate_json`` handler bound to *workspace_dir*."""

    async def handler(args: dict[str, Any]) -> dict[str, Any]:
        file_path = args.get("file_path", "")
        schema_name = args.get("schema_name", "")

        # Resolve relative to workspace
        abs_path = os.path.join(workspace_dir, file_path) if not os.path.isabs(file_path) else file_path

        if not os.path.exists(abs_path):
            return {
                "content": [{"type": "text", "text": f"File not found: {abs_path}"}],
                "is_error": True,
            }

        try:
            with open(abs_path, encoding="utf-8") as fh:
                data = json.load(fh)
        except json.JSONDecodeError as exc:
            return {
                "content": [{"type": "text", "text": f"Invalid JSON: {exc}"}],
                "is_error": True,
            }

        # If it's a list schema (feature_list), validate each item
        if schema_name == "feature_list":
            if not isinstance(data, list):
                return {
                    "content": [{"type": "text", "text": "feature_list must be a JSON array"}],
                    "is_error": True,
                }
            item_keys = _REQUIRED_KEYS.get("feature_list_item", [])
            errors = []
            for i, item in enumerate(data):
                missing = [k for k in item_keys if k not in item]
                if missing:
                    errors.append(f"Item {i}: missing keys {missing}")
            if errors:
                return {
                    "content": [{"type": "text", "text": "Validation errors:\n" + "\n".join(errors)}],
                    "is_error": True,
                }
            return {"content": [{"type": "text", "text": f"Valid feature_list with {len(data)} items."}]}

        # Single-object schemas
        required = _REQUIRED_KEYS.get(schema_name)
        if required is None:
            return {
                "content": [{"type": "text", "text": f"Unknown schema: {schema_name}. "
                             f"Available: {', '.join(_REQUIRED_KEYS.keys())}"}],
                "is_error": True,
            }

        if not isinstance(data, dict):
            return {
                "content": [{"type": "text", "text": f"{schema_name} must be a JSON object"}],
                "is_error": True,
            }

        missing = [k for k in required if k not in data]
        if missing:
            return {
                "content": [{"type": "text", "text": f"Missing required keys: {missing}"}],
                "is_error": True,
            }

        return {"content": [{"type": "text", "text": f"Valid {schema_name}."}]}

    return handler


# ---------------------------------------------------------------------------
# Public factory
# ---------------------------------------------------------------------------

def create_harness_tools(workspace_dir: str):
    """Build an in-process MCP server with harness-specific tools.

    Returns a config object suitable for ``ClaudeCodeOptions.mcp_servers["harness"]``.
    """
    update_progress_tool = SdkMcpTool(
        name="update_progress",
        description="Append a timestamped progress entry to claude-progress.txt",
        input_schema={
            "type": "object",
            "properties": {
                "phase": {
                    "type": "string",
                    "description": "Current phase (e.g. 'sprint_1_build', 'planning')",
                },
                "message": {
                    "type": "string",
                    "description": "Progress message to log",
                },
            },
            "required": ["phase", "message"],
        },
        handler=_make_update_progress(workspace_dir),
    )

    validate_json_tool = SdkMcpTool(
        name="validate_json",
        description=(
            "Validate a JSON file against a known schema. "
            "Available schemas: product_spec, feature_list, sprint_contract, qa_report, contract_review."
        ),
        input_schema={
            "type": "object",
            "properties": {
                "file_path": {
                    "type": "string",
                    "description": "Path to the JSON file (relative to workspace or absolute)",
                },
                "schema_name": {
                    "type": "string",
                    "description": "Schema to validate against",
                    "enum": list(_REQUIRED_KEYS.keys()) + ["feature_list"],
                },
            },
            "required": ["file_path", "schema_name"],
        },
        handler=_make_validate_json(workspace_dir),
    )

    return create_sdk_mcp_server(
        name="harness",
        version="1.0.0",
        tools=[update_progress_tool, validate_json_tool],
    )

File: src/tools.ts
/**
 * Outils MCP personnalisés exposés aux agents via un serveur in-process.
 *
 * La fabrique `createHarnessTools` retourne une config `McpSdkServerConfigWithInstance`
 * passable à `Options.mcpServers`.
 */

import { appendFileSync, existsSync, readFileSync } from "node:fs";
import { isAbsolute, join } from "node:path";

import {
  createSdkMcpServer,
  tool,
  type McpSdkServerConfigWithInstance,
} from "@anthropic-ai/claude-agent-sdk";
import { z } from "zod";

// ---------------------------------------------------------------------------
// Schémas JSON utilisés par l'outil `validate_json`
// ---------------------------------------------------------------------------

export const REQUIRED_KEYS: Record<string, string[]> = {
  product_spec: ["name", "description", "design_system", "sprints", "stack"],
  feature_list_item: ["id", "sprint", "category", "description", "passes"],
  sprint_contract: ["sprint_id", "sprint_name", "agreed_deliverables"],
  qa_report: ["sprint_id", "overall_score", "verdict", "scores", "bugs"],
  contract_review: ["sprint_id", "approved", "feedback"],
};

// ---------------------------------------------------------------------------
// Logique pure de validation (reproduction du handler validate_json Python)
// ---------------------------------------------------------------------------

export function validateJson(
  workspaceDir: string,
  filePath: string,
  schemaName: string,
): { content: { type: "text"; text: string }[]; isError?: boolean } {
  // Résolution relative au workspace
  const absPath = isAbsolute(filePath) ? filePath : join(workspaceDir, filePath);

  if (!existsSync(absPath)) {
    return {
      content: [{ type: "text", text: `File not found: ${absPath}` }],
      isError: true,
    };
  }

  let data: unknown;
  try {
    const raw = readFileSync(absPath, { encoding: "utf-8" });
    data = JSON.parse(raw);
  } catch (exc) {
    const message = exc instanceof Error ? exc.message : String(exc);
    return {
      content: [{ type: "text", text: `Invalid JSON: ${message}` }],
      isError: true,
    };
  }

  // S'il s'agit d'un schéma de liste (feature_list), valider chaque item
  if (schemaName === "feature_list") {
    if (!Array.isArray(data)) {
      return {
        content: [{ type: "text", text: "feature_list must be a JSON array" }],
        isError: true,
      };
    }
    const itemKeys = REQUIRED_KEYS["feature_list_item"] ?? [];
    const errors: string[] = [];
    data.forEach((item, i) => {
      const isObject = typeof item === "object" && item !== null && !Array.isArray(item);
      const record = isObject ? (item as Record<string, unknown>) : {};
      const missing = itemKeys.filter((k) => !(k in record));
      if (missing.length > 0) {
        errors.push(`Item ${i}: missing keys [${missing.map((k) => `'${k}'`).join(", ")}]`);
      }
    });
    if (errors.length > 0) {
      return {
        content: [{ type: "text", text: "Validation errors:\n" + errors.join("\n") }],
        isError: true,
      };
    }
    return {
      content: [{ type: "text", text: `Valid feature_list with ${data.length} items.` }],
    };
  }

  // Schémas objet unique
  const required = REQUIRED_KEYS[schemaName];
  if (required === undefined) {
    return {
      content: [
        {
          type: "text",
          text: `Unknown schema: ${schemaName}. Available: ${Object.keys(REQUIRED_KEYS).join(", ")}`,
        },
      ],
      isError: true,
    };
  }

  const isObject = typeof data === "object" && data !== null && !Array.isArray(data);
  if (!isObject) {
    return {
      content: [{ type: "text", text: `${schemaName} must be a JSON object` }],
      isError: true,
    };
  }

  const record = data as Record<string, unknown>;
  const missing = required.filter((k) => !(k in record));
  if (missing.length > 0) {
    return {
      content: [
        { type: "text", text: `Missing required keys: [${missing.map((k) => `'${k}'`).join(", ")}]` },
      ],
      isError: true,
    };
  }

  return { content: [{ type: "text", text: `Valid ${schemaName}.` }] };
}

// ---------------------------------------------------------------------------
// Horodatage UTC au format "%Y-%m-%d %H:%M:%S UTC" (comme Python)
// ---------------------------------------------------------------------------

function utcTimestamp(): string {
  const iso = new Date().toISOString(); // ex: 2026-05-28T19:25:07.123Z
  const datePart = iso.slice(0, 10);
  const timePart = iso.slice(11, 19);
  return `${datePart} ${timePart} UTC`;
}

// ---------------------------------------------------------------------------
// Fabrique publique
// ---------------------------------------------------------------------------

export function createHarnessTools(workspaceDir: string): McpSdkServerConfigWithInstance {
  const updateProgress = tool(
    "update_progress",
    "Append a timestamped progress entry to claude-progress.txt",
    {
      phase: z.string(),
      message: z.string(),
    },
    async ({ phase, message }) => {
      const path = join(workspaceDir, "claude-progress.txt");
      const ts = utcTimestamp();
      const line = `[${ts}] [${phase}] ${message}\n`;
      appendFileSync(path, line, { encoding: "utf-8" });
      return {
        content: [{ type: "text", text: `Progress updated: ${line.trim()}` }],
      };
    },
  );

  const validateJsonTool = tool(
    "validate_json",
    "Validate a JSON file against a known schema. " +
      "Available schemas: product_spec, feature_list, sprint_contract, qa_report, contract_review.",
    {
      file_path: z.string(),
      schema_name: z.enum(
        [...Object.keys(REQUIRED_KEYS), "feature_list"] as unknown as [string, ...string[]],
      ),
    },
    async ({ file_path, schema_name }) => validateJson(workspaceDir, file_path, schema_name),
  );

  return createSdkMcpServer({
    name: "harness",
    version: "1.0.0",
    tools: [updateProgress, validateJsonTool],
  });
}
